compiler/modules/nback/basic.bal

```diff
@@ -230,34 +230,122 @@ function buildCall(llvm:Builder builder, Scaffold scaffold, bir:CallInsn insn) r
     t:FunctionAtomicType atomic = <t:FunctionAtomicType>t:functionAtomicType(scaffold.typeContext(), insn.operands[0].semType);
     t:FunctionSignature signature = t:functionSignature(scaffold.typeContext(), atomic);
-    llvm:Value[] args = check buildFunctionCallArgs(builder, scaffold, signature.paramTypes, signature.paramTypes, from int i in 1 ..< insn.operands.length() select insn.operands[i]);
-    llvm:PointerType fnStructPtrTy = llvm:pointerType(llvm:structType([llvm:pointerType(buildFunctionSignature(signature))]));
-    llvm:PointerValue fnStructTaggedPtr = <llvm:PointerValue>builder.load(scaffold.address(insn.operands[0]));
-    llvm:Value unTaggedVal = builder.iBitwise("and",
-                                               builder.ptrToInt(fnStructTaggedPtr, LLVM_INT),
-                                               constInt(scaffold, POINTER_MASK));
-    llvm:PointerValue unTaggedPtr = builder.getElementPtr(constNil(scaffold), [unTaggedVal], "inbounds");
-    llvm:PointerValue fnStructPtr = builder.bitCast(builder.addrSpaceCast(unTaggedPtr, LLVM_TAGGED_PTR_WITHOUT_ADDR_SPACE), fnStructPtrTy);
-    llvm:PointerValue fnGlobalPtr = builder.getElementPtr(fnStructPtr, [constIndex(scaffold, 0), constIndex(scaffold, 0)], "inbounds");
+    llvm:PointerType fnStructPtrTy = llvm:pointerType(functionValueType(signature));
+    llvm:PointerValue unTaggedPtr = <llvm:PointerValue>builder.call(scaffold.getIntrinsicFunction("ptrmask.p1.i64"),
+                                                                    [<llvm:PointerValue>builder.load(scaffold.address(insn.operands[0])),
+                                                                     constInt(scaffold, POINTER_MASK)]);
+    llvm:ConstPointerValue llSignature = scaffold.getFunctionSignatureValue(signature);
+    llvm:Value isExact = buildRuntimeFunctionCall(builder, scaffold, functionIsExactFunction,
+                                                  [llSignature, builder.bitCast(unTaggedPtr, LLVM_FUNCTION_PTR)]);
+    llvm:PointerValue funcStructPtr = builder.bitCast(builder.addrSpaceCast(unTaggedPtr,
+                                                                            LLVM_TAGGED_PTR_WITHOUT_ADDR_SPACE),
+                                                      fnStructPtrTy);
+    llvm:BasicBlock ifExact = scaffold.addBasicBlock();
+    llvm:BasicBlock ifNotExact = scaffold.addBasicBlock();
+    llvm:BasicBlock afterCall = scaffold.addBasicBlock();
+    builder.condBr(isExact, ifExact, ifNotExact);
+    builder.positionAtEnd(ifExact);
+    check buildExactCall(builder, scaffold, insn, afterCall, funcStructPtr, signature);
+    builder.positionAtEnd(ifNotExact);
+    check buildInexactCall(builder, scaffold, insn, afterCall, funcStructPtr, signature);
+    builder.positionAtEnd(afterCall);
+}
+
+function buildExactCall(llvm:Builder builder, Scaffold scaffold, bir:CallIndirectInsn insn,
```

Comment:
Can we factor out some commmon functionality between buildExactCall and buildCall?

buildExactCall doesn't feel quite the right name.

Reply:
Fixed